### PR TITLE
matcher-optimizations-from-0.6.0

### PR DESCRIPTION
**Pull request by koloale**

### CODE REVIEW

File: src/main/scala/com/wavesplatform/matcher/model/OrderInfo.scala [hunk 1]
```diff
@@ -25,8 +25,8 @@ case class OrderInfo(amount: Long, filled: Long, canceled: Boolean) {
 }
 
 object OrderInfo {
-  def limitSum(x: Long, y: Long): Long = Try(math.max(0L, Math.addExact(x, y))).getOrElse(Long.MaxValue)
-  implicit val limitSemigroup: Semigroup[Long] = (x: Long, y: Long) => limitSum(x, y)
+  def safeSum(x: Long, y: Long): Long =  Try(Math.addExact(x, y)).getOrElse(Long.MaxValue)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/model/OrderInfo.scala]: These aleady exist in main code repo, please use that

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistoryStorage.scala [hunk 2]
```diff
@@ -9,7 +9,7 @@ import scorex.transaction.assets.exchange.AssetPair
 class OrderHistoryStorage(db: MVStore) {
   val orders: util.Map[String, String] = db.openMap("orders")
   val ordersInfo: util.Map[String, String] = db.openMap("ordersInfo")
-  val pairAddressToOrderIds: util.Map[AssetPairAddressKey, Set[String]] = db.openMap("pairAddressToOrderIds")
+  val pairAddressToOrderIds: util.Map[AssetPairAddressKey, Array[String]] = db.openMap("pairAddressToOrderIds")
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistoryStorage.scala]: consider using `compositeKey` if you need to access to "last 50 orders" or similar

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala [hunk 3]
```diff
@@ -96,14 +116,19 @@ case class OrderHistoryImpl(p: OrderHistoryStorage) extends OrderHistory with Sc
...
   override def openVolume(assetAcc: AssetAcc): Long = {
     val asset = assetAcc.assetId.map(_.base58).getOrElse(AssetPair.WavesName)
-    Option(p.addressToOrderPortfolio.get(assetAcc.account.address)).flatMap(_.get(asset)).getOrElse(0L)
+    Option(p.addressToOrderPortfolio.get(assetAcc.account.address)).flatMap(_.get(asset)).map(math.max(0L, _)).getOrElse(0L)
   }
 
   override def ordersByPairAndAddress(assetPair: AssetPair, address: String): Set[String] = {
     val pairAddressKey = OrderHistoryStorage.assetPairAddressKey(assetPair, address)
-    Option(p.pairAddressToOrderIds.get(pairAddressKey)).getOrElse(Set())
+    Option(p.pairAddressToOrderIds.get(pairAddressKey)).map(_.takeRight(100).toSet).getOrElse(Set())
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala]: `100` should be parameter

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala [hunk 4]
```diff
@@ -46,7 +58,7 @@ case class OrderHistoryImpl(p: OrderHistoryStorage) extends OrderHistory with Sc
 
   def saveOpenPortfolio(event: Event): Unit = {
     Events.createOpenPortfolio(event).foreach{ case(addr, op) =>
-      val prev = openPortfolio(addr)
+      val prev = Option(p.addressToOrderPortfolio.get(addr)).map(OpenPortfolio(_)).getOrElse(OpenPortfolio.empty)
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala]: just FYI, threre's  `.orEmpty` for Option[_ : Monoid]

File: src/main/scala/com/wavesplatform/matcher/market/OrderHistoryActor.scala [hunk 5]
```diff
@@ -45,27 +39,44 @@ class OrderHistoryActor(val settings: MatcherSettings, val storedState: StateRea
...
       deleteFromOrderHistory(req)
-    case GetOrderStatus(_, id) =>
+    case GetOrderStatus(_, id, _) =>
       sender() ! GetOrderStatusResponse(orderHistory.orderStatus(id))
+    case GetTradableBalance(assetPair, addr, _) =>
+      sender() ! getPairTradableBalance(assetPair, addr)
+  }
+
+  override def receive: Receive = {
+    case req: ExpirableOrderHistoryRequest =>
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/market/OrderHistoryActor.scala]: Is any of this tested?

File: src/main/scala/com/wavesplatform/matcher/MatcherSettings.scala [hunk 6]
```diff
@@ -47,12 +52,16 @@ object MatcherSettings {
...
     val maxTimestampDiff = config.as[FiniteDuration](s"$configPath.max-timestamp-diff")
 
-    val orderHistoryFile = config.as[String](s"$configPath.order-history-file")
+    val orderHistoryFile = config.getAs[File](s"$configPath.order-history-file")
+    val txHistoryFile = config.getAs[File](s"$configPath.tx-history-file")
 
-    val isMigrateToNewOrderHistoryStorage = !new File(orderHistoryFile).exists()
+    val isMigrateToNewOrderHistoryStorage = !orderHistoryFile.exists(_.exists())
+
+    val blacklistedAssets = config.as[List[String]](s"$configPath.blacklisted-assets")
```
Comment (ismagin) [inline src/main/scala/com/wavesplatform/matcher/MatcherSettings.scala]: is `config.as[Set[String]]` possible?

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala (hunk 4 above)
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala]: changed

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala (hunk 3 above)
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistory.scala]: If orders per address exceed 1000, earlier canceled orders removed from the Array, to not to allow to grows maps indefinitely. So the Array won't grow too much.

File: src/main/scala/com/wavesplatform/matcher/model/OrderHistoryStorage.scala (hunk 2 above)
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/model/OrderHistoryStorage.scala]: If orders per address exceed 1000, earlier canceled orders removed from the Array, to not to allow to grows maps indefinitely. So the Array won't grow too much.

File: src/main/scala/com/wavesplatform/matcher/model/OrderInfo.scala (hunk 1 above)
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/model/OrderInfo.scala]: For OrderOpen Portfolio I need Long.MaxValue not MinValue

**Comment by koloale** [comment]
Blacklist assets test added

File: src/main/scala/com/wavesplatform/matcher/market/OrderHistoryActor.scala (hunk 5 above)
Comment (koloale) [inline src/main/scala/com/wavesplatform/matcher/market/OrderHistoryActor.scala]: Added OrderHistoryActorSpecification

**Comment by koloale** [comment]
IT test added

**Pull request merged**